src/critnib/critnib.c

```diff
@@ -277,8 +320,22 @@ static void free_leaf(struct critnib *__restrict c,
         return;
     }
 
-    utils_atomic_store_release_ptr((void **)&k->value, c->deleted_leaf);
-    utils_atomic_store_release_ptr((void **)&c->deleted_leaf, k);
+    if (c->cb_free_leaf) {
+        uint64_t ref_count;
+        utils_atomic_load_acquire_u64(&k->ref_count, &ref_count);
+        if (ref_count == 0) {
+            utils_atomic_store_release_ptr(&k->value, c->deleted_leaf);
+            utils_atomic_store_release_ptr((void **)&c->deleted_leaf, k);
```

Comment:
This is not how corectly we update lock free single headed list. 

If two threads adds to this list concurently they can have the same "next" node

Reply:
This is the way critnib has been doing it so far and everybody have considered critnib very thread-safe so far.
How do you propose to fix it?

Reply:
Done